### PR TITLE
Blockchain pruning support for new databases

### PR DESCRIPTION
**Pull request by benevanoff**
Addresses #2087

Wallets set up in simple mode will use a pruned node by default going forward. Pruning is optional and turned off by default in advanced mode.

### CODE REVIEW

**Comment by selsta** [comment]
There is a problem here. If someone has a local daemon and starts simple mode the daemon will silently wait ages during start to prune the existing blockchain.

**Comment by benevanoff** [comment]
@selsta I've changed this PR to first check if the currently selected DB folder already exists before using the `--prune-blockchain` flag which should prevent this issue.

File: wizard/WizardDaemonSettings.qml [hunk 1]
```diff
@@ -94,6 +94,13 @@ ColumnLayout {
             }
         }
 
+        MoneroComponents.CheckBox {
+            id: pruneBlockchainCheckBox
+            checked: persistentSettings.pruneBlockchain
+            onClicked: persistentSettings.pruneBlockchain = true
```
Comment (selsta) [inline wizard/WizardDaemonSettings.qml:100]: Is it possible to uncheck it again? Seems like this checkbox can only be set to true.

**Review by selsta — COMMENTED** [review]
If you check prune blockchain during the wizard nothing will happen if a chain already exists. There should be a warning there else it might be confusing.

File: wizard/WizardDaemonSettings.qml (hunk 1 above)
Comment (benevanoff) [reply wizard/WizardDaemonSettings.qml]: Oops, thanks

**Comment by benevanoff** [comment]
@selsta thanks, updated accordingly

File: wizard/WizardDaemonSettings.qml [hunk 2]
```diff
@@ -92,6 +92,35 @@ ColumnLayout {
...
+            }
+        }
+
+        RowLayout {
+            id: pruningOptionRow
+            MoneroComponents.CheckBox {
+                id: pruneBlockchainCheckBox
+                checked: persistentSettings.pruneBlockchain
+                onClicked: {
+                    if (daemonManager.validateDataDir(persistentSettings.blockchainDataDir).lmdbExists) {
```
Comment (selsta) [inline wizard/WizardDaemonSettings.qml:106]: As far as I can see persistentSettings.blockchainDataDir is only set if a custom blockchain data dir is used, so if someone has a blockchain in the default path this does not work.

Comment (benevanoff) [reply wizard/WizardDaemonSettings.qml]: gotcha

File: wizard/WizardDaemonSettings.qml [hunk 3]
```diff
@@ -92,6 +92,35 @@ ColumnLayout {
...
+                    persistentSettings.pruneBlockchain =  !persistentSettings.pruneBlockchain
+                    this.checked = persistentSettings.pruneBlockchain
+                }
+                text: qsTr("Prune blockchain") + translationManager.emptyString
+            }
+
+            Text {
+                id: existingDbWarning
+                text: "A blockchain database already exists here. Select a new location to start a pruned node"
+                visible: daemonManager.checkLmdbExists(persistentSettings.blockchainDataDir)
```
Comment (selsta) [inline wizard/WizardDaemonSettings.qml:120]: daemonManager can be undefined / null. Same change at other places daemonManager is used.

```suggestion
                visible: daemonManager ? daemonManager.checkLmdbExists(persistentSettings.blockchainDataDir) : false
```

File: wizard/WizardDaemonSettings.qml [hunk 4]
```diff
@@ -92,6 +92,35 @@ ColumnLayout {
...
+            onTextUpdated: {
+                existingDbWarning.visible = daemonManager.checkLmdbExists(persistentSettings.blockchainDataDir)
+            }
+        }
+
+        RowLayout {
+            id: pruningOptionRow
+            MoneroComponents.CheckBox {
+                id: pruneBlockchainCheckBox
+                checked: persistentSettings.pruneBlockchain
```
Comment (selsta) [inline wizard/WizardDaemonSettings.qml:104]:
```suggestion
                checked: persistentSettings.pruneBlockchain
                enabled: !existingDbWarning.visible
```

**Review by selsta — COMMENTED** [review]
This does not work on my Mac yet, did not figure out why (will investigate later). Left some comments.

**Comment by benevanoff** [comment]
Strange, I went back and double checked with `check_blockchain_pruning` after spinning up a new DB from the create new wallet wizard in a custom set folder on my Windows machine and it works fine. You're sure you're on the most up to date revision? Some of those commits in between are broken (oops...)

File: wizard/WizardModeSelection.qml [hunk 5]
```diff
@@ -91,6 +91,7 @@ Rectangle {
 
                 onMenuClicked: {
                     if(appWindow.persistentSettings.nettype == 0){
+                        appWindow.persistentSettings.pruneBlockchain = true;
```
Comment (xiphon) [inline wizard/WizardModeSelection.qml:94]: ?

File: wizard/WizardDaemonSettings.qml [hunk 6]
```diff
@@ -92,6 +92,34 @@ ColumnLayout {
...
+                    persistentSettings.pruneBlockchain =  !persistentSettings.pruneBlockchain
+                    this.checked = persistentSettings.pruneBlockchain
+                }
+                text: qsTr("Prune blockchain") + translationManager.emptyString
+            }
+
+            Text {
+                id: existingDbWarning
+                text: "A blockchain database already exists here. Select a new location to start a pruned node"
+                visible: daemonManager ? daemonManager.checkLmdbExists(persistentSettings.blockchainDataDir) : false
```
Comment (xiphon) [inline wizard/WizardDaemonSettings.qml:119]: Should use `blockchainFolder.text`, not `persistentSettings.blockchainDataDir`

File: wizard/WizardModeSelection.qml (hunk 5 above)
Comment (benevanoff) [reply wizard/WizardModeSelection.qml:94]: Prune by default for mainnet simple mode

Comment (selsta) [reply wizard/WizardModeSelection.qml:94]: Simple mode does not sync a chain so it does not make sense to prune here.

Comment (benevanoff) [reply wizard/WizardModeSelection.qml:94]: Ah you’re right

Comment (selsta) [reply wizard/WizardModeSelection.qml:94]: So only necessary for simple mode (bootstrap)

File: wizard/WizardModeSelection.qml [hunk 7]
```diff
@@ -119,6 +119,7 @@ Rectangle {
...
```
Comment (xiphon) [inline wizard/WizardModeSelection.qml:122]: Switch from `Advanced mode` to `Simple mode (bootstrap)` and immediately switch back to `Advanced mode`. You have pruning enabled.

File: wizard/WizardDaemonSettings.qml [hunk 8]
```diff
@@ -92,6 +92,34 @@ ColumnLayout {
...
+                text: qsTr("Prune blockchain") + translationManager.emptyString
+            }
+
+            Text {
+                id: existingDbWarning
+                text: "A blockchain database already exists here. Select a new location to start a pruned node"
+                visible: daemonManager ? daemonManager.checkLmdbExists(blockchainFolder.text) : false
+                color: MoneroComponents.Style.defaultFontColor
+                font.family: MoneroComponents.Style.fontRegular.name
+            }
```
Comment (xiphon) [inline wizard/WizardDaemonSettings.qml:122]: Showing the warning using `appWindow.showStatusMessage` on `pruneBlockchainCheckBox` click will probably look better.

Comment (benevanoff) [reply wizard/WizardDaemonSettings.qml:122]: mmmm i don't think so. I tried that and it honestly looked horrible. plus then you have to enable the check box even if you dont really want the user to click it

File: wizard/WizardModeSelection.qml (hunk 7 above)
Comment (benevanoff) [reply wizard/WizardModeSelection.qml:122]: got it

**Review by xiphon — APPROVED** [review]

File: wizard/WizardDaemonSettings.qml [hunk 9]
```diff
@@ -92,6 +92,34 @@ ColumnLayout {
                     blockchainFolder.focus = true;
                 }
             }
+            onTextUpdated: {
+                var dbExists = daemonManager ? daemonManager.checkLmdbExists(blockchainFolder.text) : false
+                existingDbWarning.visible = dbExists
+                if (dbExists)
+                    pruneBlockchainCheckBox.checked = false;
+            }
```
Comment (selsta) [inline wizard/WizardDaemonSettings.qml:100]:
```suggestion
```

Seems redundant, works without this code here too.

existingDbWarning.visible should get updated once blockchainFolder.text changes, and pruneBlockchainCheckBox.checked also gets updated once existingDbWarning.visible changes.

**Comment by selsta** [comment]
Ok, don’t know if it was one of the recent changes but I got it to work on Mac too.

**Review by selsta — APPROVED** [review]

**Pull request merged**